### PR TITLE
Create handler

### PR DESCRIPTION
**Pull request by sfbaker7**
*Description of changes:*

Adds create handler + tests

*Testing*
Unit + Integ 

```
[INFO] -------------------------------------------------------
[INFO]  T E S T S
[INFO] -------------------------------------------------------
[INFO] Running software.amazon.logs.querydefinitions.ReadHandlerTest
[INFO] Tests run: 4, Failures: 0, Errors: 0, Skipped: 0, Time elapsed: 4.102 s - in software.amazon.logs.querydefinitions.ReadHandlerTest
[INFO] Running software.amazon.logs.querydefinitions.CreateHandlerTest
[INFO] Tests run: 4, Failures: 0, Errors: 0, Skipped: 0, Time elapsed: 0.083 s - in software.amazon.logs.querydefinitions.CreateHandlerTest
[INFO] Running software.amazon.logs.querydefinitions.ListHandlerTest
[INFO] Tests run: 4, Failures: 0, Errors: 0, Skipped: 0, Time elapsed: 0.067 s - in software.amazon.logs.querydefinitions.ListHandlerTest
[INFO] Running software.amazon.logs.querydefinitions.UpdateHandlerTest
[INFO] Tests run: 4, Failures: 0, Errors: 0, Skipped: 0, Time elapsed: 0.044 s - in software.amazon.logs.querydefinitions.UpdateHandlerTest
[INFO] Running software.amazon.logs.querydefinitions.DeleteHandlerTest
[INFO] Tests run: 5, Failures: 0, Errors: 0, Skipped: 0, Time elapsed: 0.13 s - in software.amazon.logs.querydefinitions.DeleteHandlerTest
[INFO]
```
By submitting this pull request, I confirm that you can use, modify, copy, and redistribute this contribution, under the terms of your choice.

### CODE REVIEW

File: aws-logs-querydefinitions/src/main/java/software/amazon/logs/querydefinitions/CreateHandler.java [hunk 1]
```diff
@@ -16,8 +19,16 @@
         final Logger logger) {
 
         final ResourceModel model = request.getDesiredResourceState();
```
Comment (anugarg07) [inline aws-logs-querydefinitions/src/main/java/software/amazon/logs/querydefinitions/CreateHandler.java:21]: Is this the one to perform validation against required fields?

File: aws-logs-querydefinitions/src/main/java/software/amazon/logs/querydefinitions/CreateHandler.java [hunk 2]
```diff
@@ -16,8 +19,16 @@
...
+        PutQueryDefinitionResponse putQueryDefinitionResponse;
+        try {
+            putQueryDefinitionResponse = proxy.injectCredentialsAndInvokeV2(Translator.translateToCreateRequest(model), ClientBuilder.getLogsClient()::putQueryDefinition);
+        } catch (InvalidParameterException ex) {
+            throw new CfnInvalidRequestException(ResourceModel.TYPE_NAME, ex);
```
Comment (anugarg07) [inline aws-logs-querydefinitions/src/main/java/software/amazon/logs/querydefinitions/CreateHandler.java:26]: Is `CfnInvalidRequestException` handled as validation exception and exception message shown to customer on CFN side ?

File: aws-logs-querydefinitions/src/test/java/software/amazon/logs/querydefinitions/CreateHandlerTest.java [hunk 3]
```diff
@@ -12,43 +18,92 @@
...
-        assertThat(response.getResourceModel()).isEqualTo(request.getDesiredResourceState());
-        assertThat(response.getResourceModels()).isNull();
-        assertThat(response.getMessage()).isNull();
-        assertThat(response.getErrorCode()).isNull();
+                .desiredResourceState(model)
+                .build();
+
+        doThrow(InvalidParameterException.builder().message("test Error").build())
+                .when(proxy)
+                .injectCredentialsAndInvokeV2(ArgumentMatchers.any(), ArgumentMatchers.any());
```
Comment (anugarg07) [inline aws-logs-querydefinitions/src/test/java/software/amazon/logs/querydefinitions/CreateHandlerTest.java:104]: Can we add UTs for :

-     when CWL throws some validation exception i.e 400 status code
-     when CWL throws a 500

File: aws-logs-querydefinitions/src/main/java/software/amazon/logs/querydefinitions/CreateHandler.java [hunk 4]
```diff
@@ -16,8 +19,16 @@
         final Logger logger) {
 
         final ResourceModel model = request.getDesiredResourceState();
+        PutQueryDefinitionResponse putQueryDefinitionResponse;
+        try {
+            putQueryDefinitionResponse = proxy.injectCredentialsAndInvokeV2(Translator.translateToCreateRequest(model), ClientBuilder.getLogsClient()::putQueryDefinition);
+        } catch (InvalidParameterException ex) {
```
Comment (anugarg07) [inline aws-logs-querydefinitions/src/main/java/software/amazon/logs/querydefinitions/CreateHandler.java:25]: Are there any retriable exceptions we want to handle here?

File: aws-logs-querydefinitions/src/test/java/software/amazon/logs/querydefinitions/CreateHandlerTest.java (hunk 3 above)
Comment (anugarg07) [reply aws-logs-querydefinitions/src/test/java/software/amazon/logs/querydefinitions/CreateHandlerTest.java:104]: > Can we add UTs for :
> 
>     * when CWL throws some validation exception i.e 400 status code
> 
>     * when CWL throws a 500
> 
> 
> Assert that the returned response contains correct exception type. Depending upon exception type, CFN might be deciding if/not to show error msg to customers.

Seems there is `Handler` code associated with each exception type https://github.com/aws-cloudformation/cloudformation-cli-java-plugin/blob/master/src/main/java/software/amazon/cloudformation/exceptions/CfnInvalidRequestException.java and CFN would be using it to determine if to rety/not

**Review by anugarg07 — COMMENTED** [review]
Some comments around failure handling

File: aws-logs-querydefinitions/src/main/java/software/amazon/logs/querydefinitions/CreateHandler.java (hunk 1 above)
Comment (sfbaker7) [reply aws-logs-querydefinitions/src/main/java/software/amazon/logs/querydefinitions/CreateHandler.java:21]: doesn't seem like cfn does field validation from what I can tell. the required fields check would fall onto the Logs backend when we send the request

File: aws-logs-querydefinitions/src/main/java/software/amazon/logs/querydefinitions/CreateHandler.java (hunk 2 above)
Comment (sfbaker7) [reply aws-logs-querydefinitions/src/main/java/software/amazon/logs/querydefinitions/CreateHandler.java:26]: yeah they will get this exception in the case that they don't put a required field

File: aws-logs-querydefinitions/src/main/java/software/amazon/logs/querydefinitions/CreateHandler.java (hunk 4 above)
Comment (sfbaker7) [reply aws-logs-querydefinitions/src/main/java/software/amazon/logs/querydefinitions/CreateHandler.java:25]: I think we should only retry the default SDK retryable errors

File: aws-logs-querydefinitions/src/test/java/software/amazon/logs/querydefinitions/CreateHandlerTest.java (hunk 3 above)
Comment (sfbaker7) [reply aws-logs-querydefinitions/src/test/java/software/amazon/logs/querydefinitions/CreateHandlerTest.java:104]: I will add two more test case, one for not found and another for service internal error. Besides that, I think we should either let the SDK retry (with the 500s) or let cfn handle.

Comment (sfbaker7) [reply aws-logs-querydefinitions/src/test/java/software/amazon/logs/querydefinitions/CreateHandlerTest.java:102]: will also add test to make sure that the correct exception is thrown

File: aws-logs-querydefinitions/src/main/java/software/amazon/logs/querydefinitions/CreateHandler.java (hunk 1 above)
Comment (anugarg07) [reply aws-logs-querydefinitions/src/main/java/software/amazon/logs/querydefinitions/CreateHandler.java:25]: I figured it is done in lambda framework code , before CRUD handlers are invoked  - https://github.com/aws-cloudformation/cloudformation-cli-java-plugin/blob/master/src/main/java/software/amazon/cloudformation/AbstractWrapper.java#L199 . So we are good.

You can test the validation cases when doing integration testing using CFN templates.

File: aws-logs-querydefinitions/src/main/java/software/amazon/logs/querydefinitions/CreateHandler.java (hunk 2 above)
Comment (anugarg07) [reply aws-logs-querydefinitions/src/main/java/software/amazon/logs/querydefinitions/CreateHandler.java:30]: See comment above ; this validation is already done in base lambda code.

File: aws-logs-querydefinitions/src/main/java/software/amazon/logs/querydefinitions/CreateHandler.java (hunk 4 above)
Comment (anugarg07) [reply aws-logs-querydefinitions/src/main/java/software/amazon/logs/querydefinitions/CreateHandler.java:29]: agree

**Pull request merged**